src/server/gc_worker/gc_worker.rs

```diff
@@ -422,6 +454,139 @@ where
         Ok((handled_keys, wasted_keys))
     }
 
+    fn raw_gc_keys(
+        &mut self,
+        keys: Vec<Key>,
+        safe_point: TimeStamp,
+        regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
+    ) -> Result<(usize, usize)> {
+        let first_key = keys.first().unwrap();
+        let first_raw_key = first_key.clone().append_ts(TimeStamp::max());
+        let range_start_key = first_raw_key.as_encoded();
+
+        let last_key = keys.last().unwrap();
+        let last_raw_key = last_key.clone().append_ts(TimeStamp::zero());
+        let range_end_key = last_raw_key.as_encoded();
+
+        //--------------------------------KeysInRegions-----------------------------
+        struct KeysInRegions<R: Iterator<Item = Region>> {
+            keys: Peekable<IntoIter<Key>>,
+            regions: Peekable<R>,
+        }
+        impl<R: Iterator<Item = Region>> Iterator for KeysInRegions<R> {
+            type Item = Key;
+            fn next(&mut self) -> Option<Key> {
+                loop {
+                    let region = self.regions.peek()?;
+                    let key = self.keys.peek()?.as_encoded().as_slice();
+                    if key < region.get_start_key() {
+                        self.keys.next();
+                    } else if region.get_end_key().is_empty() || key < region.get_end_key() {
+                        return self.keys.next();
+                    } else {
+                        self.regions.next();
+                    }
+                }
+            }
+        }
+
+        fn get_keys_in_regions(
+            keys: Vec<Key>,
+            regions_provider: Option<(u64, Arc<dyn RegionInfoProvider>)>,
+        ) -> Result<Box<dyn Iterator<Item = Key>>> {
+            if keys.len() >= 2 {
+                if let Some((store_id, region_info_provider)) = regions_provider {
+                    let start = keys.first().unwrap().as_encoded();
+                    let end = keys.last().unwrap().as_encoded();
+                    let regions = box_try!(region_info_provider.get_regions_in_range(start, end))
+                        .into_iter()
+                        .filter(move |r| find_peer(r, store_id).is_some())
+                        .peekable();
+
+                    let keys = keys.into_iter().peekable();
+                    return Ok(Box::new(KeysInRegions { keys, regions }));
+                }
+            }
+            Ok(Box::new(keys.into_iter()))
+        }
+
+        let mut keys = get_keys_in_regions(keys, regions_provider)?;
+        //--------------------------------KeysInRegions-----------------------------
+
+        let mut snapshot = self
+            .engine
+            .snapshot_on_kv_engine(range_start_key, range_end_key)?;
```

Comment:
The `range_start_key` and `range_end_key` are used to create a mock region info for setting the bound of scanning. In normal cases the bounds of a region should be keys without timestamp suffix (i.e. encoded user key). Perhaps it works just fine now, I still suggest you to keep its normal usage here to prevent potential problems.

To do this, the start_key (inclusive) should be the first key itself, and the end_key (exclusive) should be the *next* key to the last key. The *next* key means the lexicographically smallest key that's greater than a specified key, which can be obtained by simply appending a `'0'` byte. Since we are using encoded form, the correct way to get it is to decode it, append zero, and encode it back. This is what line 387-395 in `gc_keys` does, which I think you can refer to.